### PR TITLE
Major fixes to store migration

### PR DESCRIPTION
**Pull request by jlarfors**
Wrote a more comprehensive test suite for testing the migrations.
Both the schema updates/changes which are handled by the store, and also the apply-ing of
those by the postgres provider.

Also added a test package, which I originally built in bubbly-auth and thought it could
be very useful in bubbly too. Would be good to standardise on this for the use of dockertest+postgres

### CODE REVIEW

File: test/dockertest.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+	require.NoErrorf(t, err, "failed to create dockertest pool")
+
+	resource, err := pool.RunWithOptions(
+		&dockertest.RunOptions{
+			Repository: "postgres",
+			Tag:        "13.0",
+			Env: []string{
+				"POSTGRES_USER=postgres",
+				"POSTGRES_PASSWORD=postgres",
+				"POSTGRES_DB=bubbly",
```
Comment (olliefr) [inline test/dockertest.go:23]: I feel these must come from the environment variables: `Tag`, `POSTGRES_USER`, `POSTGRES_PASSWORD`, `POSTGRES_DB`.

This would give us:

✔️ More flexibility
✔️ Better upgradeability
✔️ Better security

If you are _completely_ against the idea of having to set a few environment variables before running tests, 

👺 firstly, this; and
❓  at least let's consider reading the values from environmental variables while providing fall-back default values, such as the ones currently hard-coded, in case the environment variables are not set.

File: test/dockertest.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+	err = waitUntilDatabaseIsReady(t, pool, resource)
+	require.NoErrorf(t, err, "error waiting for database to be ready")
+
+	return resource
+}
+
+func waitUntilDatabaseIsReady(t *testing.T, pool *dockertest.Pool, resource *dockertest.Resource) error {
+
+	pgConnStr := fmt.Sprintf("postgresql://%s:%s@localhost:%s/%s?sslmode=disable",
+		"postgres", "postgres", resource.GetPort("5432/tcp"), "bubbly")
```
Comment (olliefr) [inline test/dockertest.go:44]: The string literals here have the same content as in the lines 19 to 23, which means we duplicate information.

I would:
✔️ Read this information from the environment variables. Or at least deduplicate it.
❓  Consider using keyword/value connection string format which is equivalent but more readable.

[Postgres 33.1.1. Connection Strings](https://www.postgresql.org/docs/13/libpq-connect.html#LIBPQ-CONNSTRING)

File: store/postgres_migration.go [hunk 3]
```diff
@@ -27,7 +28,11 @@ func psqlGenerateMigration(provider config.StoreProviderType, tenant string, ch
 			case fieldType:
 				m = append(m, "ALTER TABLE IF EXISTS "+psqlAbsTableName(tenant, tableName)+" DROP COLUMN IF EXISTS "+change.TableInfo.ElementName)
```
Comment (olliefr) [inline store/postgres_migration.go:29]: ❓ Is this... a fragment of (SQL injection) attack surface? 
❓ Do `tenant` and `tableName` come already sanitized?

File: store/postgres_migration.go [hunk 4]
```diff
@@ -204,11 +187,16 @@ func createFieldStatement(tenant string, info tableInfo, fieldInterface interfac
...
 	join, ok := (joinInterface).(core.TableJoin)
 	if !ok {
-		return "", fmt.Errorf("uniqueInterface not assignable to bool: %s", joinInterface)
+		return "", fmt.Errorf("cannot assign type %s to core.TableJoin", reflect.TypeOf(joinInterface).String())
 	}
-	if join.Single {
-		return "ALTER TABLE IF EXISTS " + psqlAbsTableName(tenant, info.TableName) + " ADD CONSTRAINT " + info.TableName + "_" + info.ElementName + "_unique UNIQUE (" + info.ElementName + tableJoinSuffix + ");", nil
-	} else {
-		return "ALTER TABLE IF EXISTS " + psqlAbsTableName(tenant, info.TableName) + " ADD COLUMN IF NOT EXISTS " + info.TableName + tableJoinSuffix + " SERIAL;", nil
+	return "ALTER TABLE IF EXISTS " + psqlAbsTableName(tenant, info.TableName) + " ADD COLUMN IF NOT EXISTS " + join.Table + tableJoinSuffix + " SERIAL;", nil
```
Comment (olliefr) [inline store/postgres_migration.go:192]: Same question regarding potential SQL injection and input sanitisation.

File: store/postgres_migration.go [hunk 5]
```diff
@@ -204,11 +187,16 @@ func createFieldStatement(tenant string, info tableInfo, fieldInterface interfac
...
+}
+
+// removeJoinStatement removes a the column which acts as a join for a table
+func removeJoinStatement(tenant string, info tableInfo, joinInterface interface{}) (string, error) {
+	join, ok := (joinInterface).(core.TableJoin)
+	if !ok {
+		return "", fmt.Errorf("cannot assign type %s to core.TableJoin", reflect.TypeOf(joinInterface).String())
 	}
+	return "ALTER TABLE IF EXISTS " + psqlAbsTableName(tenant, info.TableName) + " DROP COLUMN IF EXISTS " + join.Table + tableJoinSuffix, nil
```
Comment (olliefr) [inline store/postgres_migration.go:201]: Same question regarding potential SQL injection and input sanitisation.

File: store/postgres_migration.go [hunk 6]
```diff
@@ -204,11 +187,16 @@ func createFieldStatement(tenant string, info tableInfo, fieldInterface interfac
...
-	} else {
-		return "ALTER TABLE IF EXISTS " + psqlAbsTableName(tenant, info.TableName) + " ADD COLUMN IF NOT EXISTS " + info.TableName + tableJoinSuffix + " SERIAL;", nil
+	return "ALTER TABLE IF EXISTS " + psqlAbsTableName(tenant, info.TableName) + " ADD COLUMN IF NOT EXISTS " + join.Table + tableJoinSuffix + " SERIAL;", nil
+}
+
+// removeJoinStatement removes a the column which acts as a join for a table
+func removeJoinStatement(tenant string, info tableInfo, joinInterface interface{}) (string, error) {
+	join, ok := (joinInterface).(core.TableJoin)
+	if !ok {
+		return "", fmt.Errorf("cannot assign type %s to core.TableJoin", reflect.TypeOf(joinInterface).String())
```
Comment (olliefr) [inline store/postgres_migration.go:199]: ❓ The canonical logging format would be `"cannot assign type to core.TableJoin: %s"`, no?

File: store/postgres_migration.go [hunk 7]
```diff
@@ -204,11 +187,16 @@ func createFieldStatement(tenant string, info tableInfo, fieldInterface interfac
 func createJoinStatement(tenant string, info tableInfo, joinInterface interface{}) (string, error) {
 	join, ok := (joinInterface).(core.TableJoin)
 	if !ok {
-		return "", fmt.Errorf("uniqueInterface not assignable to bool: %s", joinInterface)
+		return "", fmt.Errorf("cannot assign type %s to core.TableJoin", reflect.TypeOf(joinInterface).String())
```
Comment (olliefr) [inline store/postgres_migration.go:190]: ❓ The canonical logging format would be `"cannot assign type to core.TableJoin: %s"`, no?

File: store/postgres_migration.go [hunk 8]
```diff
@@ -157,30 +159,11 @@ func alterColumnStatement(provider config.StoreProviderType, tenant string, info
...
-	}
-	return statements, nil
-}
-
 // this will create a column in a table, and then if specified add the UNIQUE constraint
 func createFieldStatement(tenant string, info tableInfo, fieldInterface interface{}) ([]string, error) {
 	field, ok := (fieldInterface).(core.TableField)
 	if !ok {
-		return nil, fmt.Errorf("fieldInterface not assignable to core.TableField: %s", fieldInterface)
+		return nil, fmt.Errorf("cannot assign type %s to core.TableField", reflect.TypeOf(fieldInterface).String())
```
Comment (olliefr) [inline store/postgres_migration.go:166]: ❓ The canonical logging format would be `"cannot assign type to core.TableJoin: %s"`, no?

File: store/postgres_migration.go [hunk 9]
```diff
@@ -121,7 +123,7 @@ func psqlMigrate(conn *pgxpool.Pool, tenant string, schema *bubblySchema, migr m
 func alterColumnStatement(provider config.StoreProviderType, tenant string, info tableInfo, columnType interface{}) ([]string, error) {
 	t, ok := (columnType).(cty.Type)
 	if !ok {
-		return nil, fmt.Errorf("not able to parse value to cty.Type: %s", columnType)
+		return nil, fmt.Errorf("cannot assign type %s to cty.Type", reflect.TypeOf(columnType).String())
```
Comment (olliefr) [inline store/postgres_migration.go:126]: ❓ The canonical logging format would be `"cannot assign type to core.TableJoin: %s"`, no?

File: store/postgres_migration_test.go [hunk 10]
```diff
@@ -1,151 +1,147 @@
...
 )
 
 const (
 	postgresDatabase = "bubbly"
 	postgresUser     = "postgres"
 	postgresPassword = "secret"
-
-	cockroachDatabase = "defaultdb"
-	cockroachUser     = "root"
-	cockroachPassword = "admin"
```
Comment (olliefr) [inline store/postgres_migration_test.go:27]: ❓ Is this... the time we pull the plug on (now dysfunctional) CockroachDB support?
✔️ I would delete `store/cockroachdb.go`

File: store/postgres_migration_test.go [hunk 11]
```diff
@@ -1,151 +1,147 @@
...
-	"github.com/valocode/bubbly/api/core"
-
-	"github.com/stretchr/testify/require"
+	"github.com/valocode/bubbly/test"
 )
 
 const (
 	postgresDatabase = "bubbly"
 	postgresUser     = "postgres"
 	postgresPassword = "secret"
```
Comment (olliefr) [inline store/postgres_migration_test.go:19]: ❓ These probably are not necessary since they are duplicated in the `test/dockertest.go`?

File: store/postgres_migration_test.go [hunk 12]
```diff
@@ -1,151 +1,147 @@
...
 
-	"github.com/valocode/bubbly/config"
 	"github.com/valocode/bubbly/env"
-
-	"github.com/ory/dockertest/v3"
-	"github.com/ory/dockertest/v3/docker"
-	"github.com/valocode/bubbly/api/core"
-
-	"github.com/stretchr/testify/require"
+	"github.com/valocode/bubbly/test"
```
Comment (olliefr) [inline store/postgres_migration_test.go:13]: ❔ This is more of a question, but I think both of our beloved IDEs have an option to group "our" modules together and separate from "theirs". Do we want that?

**Review by olliefr — CHANGES_REQUESTED** [review]
Wow, that's a lot of tests. Much appreciated 👍 
I've noted just a few things for your consideration...

File: test/dockertest.go (hunk 1 above)
Comment (jlarfors) [reply test/dockertest.go:23]: Could we just use the values from `bCtx` the Bubbly Context? As those come from environment variables...
What do you think?

File: test/dockertest.go (hunk 2 above)
Comment (jlarfors) [reply test/dockertest.go:44]: Yup, because everything was hard coded with consts and I got tired of updating the Bubbly Context every time you want to do something with a DB. So I hardcoded them as literal strings to the defaults :)

Reading from env vars, or better yet Bubbly Context, would solve all this

File: store/postgres_migration.go (hunk 3 above)
Comment (jlarfors) [reply store/postgres_migration.go:29]: There will be naming conventions for these. `tenant` and `tableName` will both be something like `[A-Za-b0-9-_]`

Comment (jlarfors) [reply store/postgres_migration.go:29]: The validation of names has yet to be added...

File: store/postgres_migration_test.go (hunk 10 above)
Comment (jlarfors) [reply store/postgres_migration_test.go:27]: I think we can do that yes. After this PR. The tests in here were being skipped already anyway...

File: store/postgres_migration_test.go (hunk 11 above)
Comment (jlarfors) [reply store/postgres_migration_test.go:19]: Well they are being used in `store/store_test.go`... This was a real PITA yesterday because the tests are coupled to each other. Again, `bCtx` as the source of truth would be great!

**Review by benmarsden — APPROVED** [review]
Looks excellent. `schemaDiffTests`, with a readable list of `input` and `expected` is excellent, and what we should almost always be doing, as they build our own confidence in the implementation and make it much easier to add further edge cases that we uncover later on.

File: test/dockertest.go (hunk 1 above)
Comment (olliefr) [reply test/dockertest.go:23]: Yes, sounds good to me!

**Review by olliefr — APPROVED** [review]
I'm approving this PR, based on @jlarfors responses above. I agree that the small issues I had raised can be handled in separate PRs, if the need be.
